### ISSUE TITLE
Error The File Input is Damaged or Modified

### ISSUE THREAD

**Issue by kriegs-gaster** [Picocrypt/Picocrypt#12]
I Encrypted a 30GB File using picocrypt 1.3.3 (Windows). where I used the following option to encrypt
1. Paranoid Mode
2. Deniability 

The File was Encrypted Successfully and I saved output as name.7z.pcv (for example).  

Today when i tried to decrypt. the deniability has been removed and started to decrypt. after few hours (since it large file) it reach 100% and suddenly it throws the error: The File Input is damaged or modified. The Force Decrypt option is disable by default.

So  i am stuck with unable to decrypt the files. i tried to check similar issues, but no avail.

**Comment by kriegs-gaster** [comment]
Before Starting to Decrypt the File. 
![image](https://github.com/Picocrypt/Picocrypt/assets/173260990/eae2d86b-8ba6-43ce-baf6-70e2f436f7da)

After Reaching 100%
![image](https://github.com/Picocrypt/Picocrypt/assets/173260990/792bc76c-b531-470a-8248-44676f8f94d6)

**Comment by HACKERALERT** [comment]
This is an interesting case. Are you sure the file isn't corrupted and you're using the correct password?

There's no way currently to only remove the deniability layer. I will try write a small tool to do that and possibly put it in the Extras repo. I will let you know if I have the time and manage to do it. With that, you can remove the deniability layer and with the regular volume, toggle the "force decrypt". If that succeeds, you probably had some bits corrupt.

**Comment by HACKERALERT** [comment]
This is an interesting case. Are you sure the file isn't corrupted and you're using the correct password?

**Comment by kriegs-gaster** [comment]
Yes I certain, because I usually keep hash value of original file and pico-encrypted file. before decrypt I check the hash value of the file and it was same. therefore I am sure the file isn't corrupted and I used correct password also. if I used wrong password it will give a popup saying "Incorrect password" after trying to remove Deniability.  

One thing I am not sure about is whether the file was corrupted during the encryption process. Since I have only the pico-encrypted hash and the non-encrypted hash file, if it was corrupted during the process, then the pico-encrypted hash is useless. I’m assuming that the file could be corrupted during the process. Also, I didn’t decrypt to see if the files are working fine, since they were working fine with other files (both small and large files ranging from 10 to 25GB).

**Comment by HACKERALERT** [comment]
Okay, try extracting the undeniable volume manually. Decrypt as usual and wait for the deniability removal to finish. Once decryption starts, you should see a temporary .pcv. Open task manager and kill Picocrypt (don't cancel in the app, kill in TM). Then the .pcv that didn't get cleaned up since you killed the app is the volume without deniability. Drop that into Pico and try to decrypt with the force option on. I suggest you try with a 1-3gb file first to see what temporary .pcv i'm talking about (watch file explorer carefully) before going to do it with your large volume.

**Comment by HACKERALERT** [comment]
In other words, have your file explorer open to the folder where the volume is contained. Have that window on your left, have Pico on your right. Start decrypting the deniable volume and watch as a .pcv is created. Wait until after key derivation and decryption starts before you kill the app. The temporary .pcv created is what you want.

**Comment by kriegs-gaster** [comment]
Okay, Here is the updated, I tested the decryption process with files of various sizes: 500 MB, 2 GB, and 20 GB. It worked successfully for these files.  However, when I attempted to decrypt a 30 GB original file, the same error occurred, stating, “The File Input is Damaged or Modified.”

To investigate further, I conducted an experiment. I began encrypting a 1 GB file and abruptly shut down my laptop. As a result, the temporary file was automatically saved as an encrypted file (from “file.7z.pcv.temp” to “file.7z.pcv”). This renaming of the file led to confusion, making me believe that the file had been encrypted.

**Comment by HACKERALERT** [comment]
So no files were lost, then right?

I'm not sure why the auto renaming would happen, maybe it's a system thing.

**Comment by kriegs-gaster** [comment]
I am really sorry for miscommunication since i am new. you are right there is no auto renaming but i got confused with it.
but i know how to reproduced this error.

Here is the video how to reproduced this error https://youtu.be/7LrS9ZB0038 . 
Here i test the same without deniability - https://youtu.be/DD7ErV-gQV0 .

When the software is terminated, the file is corrupted and users did not know if the file is corrupted or not. only they see .pcv at the end to believed that file is encrypted and delete the original and other files without realizing it. 

Is it possible to have a checksum for both the original file and the pico-encrypted file, as well as tamper-proof methods to ensure that the original file contents are successfully encrypted and the encrypted file has not been corrupted due to a pico termination? Is it possible to inform users that the previous file encryption was unsuccessful.

**Comment by HACKERALERT** [comment]
Thank you for the video, you are correct, there is no auto renaming, just an incomplete .pcv. I don't think there is a graceful way to solve this because the point of Task Manager killing an app is to end it unconditionally... One possible solution is to add the expected file size into the volume header so that upon decryption, if the volume is incomplete, the file sizes wouldn't match and we would know. But that would require changes to the header, which would likely break compatibility. Not sure what is best to do at the moment.

**Comment by kriegs-gaster** [comment]
> One possible solution is to add the expected file size into the volume header so that upon decryption, if the volume is incomplete, the file sizes wouldn't match and we would know. But that would require changes to the header, which would likely break compatibility. Not sure what is best to do at the moment

Although its good idea, but I don't think it would solved the problem. since picocrypt start fresh instances, it can't check incomplete .pcv (both with or without deniability). i think we should have some kind of error log tracking feature. it keeps a log all operations with minimum detailed with no tractability. one idea is if the file is encrypted successfully we can delete / shred the log file. if the software is terminated and when the new instances picks the log. it alert the users something is went wrong and ask them to encrypt the file. 


 > But that would require changes to the header, which would likely break compatibility

Can i know you are talking about forward or backward compatibility?. in case of backward compatibility i think we can do it by adding verification mark / File size at end of file. the file structure (File Header + Encrypted Content + File Size). Once the Decryption is done we can remove it. 

Kindly excuse me since i am end user and i have no technical / developer knowledge and correct me if i am wrong.

**Comment by hakavlad** [comment]
>I don't think there is a graceful way

Just don't give output files the `.pcv` extension if they are not what the user expects. Is this not graceful? Just rename the final file to `.pcv` only if it was written successfully.

For example:

input file -> encrypt to `.pcv1_tmp` -> rename to `.pcv

With Deniability:

input file -> encrypt to `.pcv1_tmp` -> reencrypt to `.pcv2_tmp` (add Deniability) -> rename to `.pcv`

Optional: check the expected file size before renaming.

**Comment by HACKERALERT** [comment]
Yes, good idea, I like it.

**Comment by njhuffman** [comment]
I like the idea of the header being empty or similar until the end, and using that as a sign that encryption terminated early. That's already how the auth tag is handled. The file renaming feels worse to me because I don't like that it could leave extra files on the system if pico is terminated at the wrong time. To be clear, that's already true of the current implementation, so it wouldn't be a new thing.

I guess the question is, if encryption is killed early, is it better to have `foo.txt.pcv` exist but be incomplete, or to have a different file. My personal preference would be to have the correctly named file, and verify myself before deleting. I'd use the builtin delete when done feature, and I would know something went wrong because the built in feature didn't delete my original file. But both options are valid, could be a setting if there are enough people on both sides.

**Comment by kriegs-gaster** [comment]
i am also thinking of same that renaming is viable option. we have to think from end user who has no technical knowledge how things work and he knows if you put the file it will be encrypted with .pcv format and vice-versa. if the encryption is terminated and has a incomplete `.pcv_temp` then user will become aware that file is not fully encrypted and start encrypting process again. 

>Optional: check the expected file size before renaming

I Recommended this option, because i forgot to check the file size of both original and encrypted which resulted this incomplete corrupted pcv file. if i check the file size during i could have avoided it.  

>  is it better to have `foo.txt.pcv` exist but be incomplete

if the user in urgent they forgot to see the file size or do verification assuming that it works. giving false belief and delete the original files. later when it decryption fails by saying "The File Input is Damaged or Modified" even with correct password it destroys the goodwill of the program. if user knows before that encryption failed then it could been avoided. 

 > correctly named file, and verify myself before deleting

generally I do manually decrypted my self and verify the files are encrypted or decrypted correctly and have a hash value of both original file and  encrypted file (due to deniability). in this case i did not do it, because 30GB Encryption takes 1 Hours to Encrypted and Another 1 Hour to Decrypt, in Total 2 Hours in verification. But automating the option is good. like verifying that file is encrypted correctly or not. 

> builtin delete when done feature and I would know something went wrong

That good way, the problem is (in my case) sometimes but rarely i accidently put wrong password or mistype it and if the orginal file is deleted, there is no way of decrypting it. that the reason i never use built-in delete feature for encryption.

**Comment by HACKERALERT** [comment]
For the existing code, it may be best to rename in progress volumes to encrypted.zip.pcv.incomplete or . unfinished. then that should be obvious to the average user.

Since @njhuffman is doing a redesign of the code, I trust that you will know what is best and you can go with what you prefer.

**Comment by HACKERALERT** [comment]
I just realized now that a section of the header is zeroed as a placeholder until encryption finishes and the MAC tag determined is written to there. At byte 309 or something like that. So it would be possible to check if that section is all zeros...

**Comment by HACKERALERT** [comment]
@kriegs-gaster @CodeCracker-oss @hakavlad 

https://github.com/Picocrypt/Picocrypt/pull/108

In this PR, I pushed a commit that adds ".incomplete" and only removes it when encryption finish. Can you help me test different options/advanced options/configurations to make sure nothing broke? (Please continue this thread in the PR, not this thread).
